File: LinkedList/dingyilianbiao.java
package LinkedList;

//单链表
class dingyilianbiao {

    class ListNode {
        int val;
        ListNode next;
        ListNode(int val) {
            this.val=val;
        }
    }
    //size存储链表元素的个数
    private int size;
    //注意这里记录的是虚拟头结点
    private ListNode head;

    //初始化链表
    public dingyilianbiao() {
        this.size = 0;
        this.head = new ListNode(0);
    }

    //获取第index个节点的数值，注意index是从0开始的，第0个节点就是虚拟头结点
    public int get(int index) {
        //如果index非法，返回-1
        if (index < 0 || index >= size) {
            return -1;
        }
        ListNode cur = head;
        //第0个节点是虚拟头节点，所以查找第 index+1 个节点
        for (int i = 0; i <= index; i++) {
            cur = cur.next;
        }
        return cur.val;
    }

    public void addAtHead(int val) {
        ListNode newNode = new ListNode(val);
        newNode.next = head.next;
        head.next = newNode;
        size++;

        // 在链表最前面插入一个节点，等价于在第0个元素前添加
        // addAtIndex(0, val);
    }

    
    public void addAtTail(int val) {
        ListNode newNode = new ListNode(val);
        ListNode cur = head;
        while (cur.next != null) {
            cur = cur.next;
        }
        cur.next = newNode;
        size++;

        // 在链表的最后插入一个节点，等价于在(末尾+1)个元素前添加
        // addAtIndex(size, val);
    }

    // 在第 index 个节点之前插入一个新节点，例如index为0，那么新插入的节点为链表的新头节点。
    // 如果 index 等于链表的长度，则说明是新插入的节点为链表的尾结点
    // 如果 index 大于链表的长度，则返回空
    public void addAtIndex(int index, int val) {
        if (index < 0 || index > size) {
            return;
        }

        //找到要插入节点的前驱
        ListNode pre = head;
        for (int i = 0; i < index; i++) {
            pre = pre.next;
        }
        ListNode newNode = new ListNode(val);
        newNode.next = pre.next;
        pre.next = newNode;
        size++;
    }

    public void deleteAtIndex(int index) {
        if (index < 0 || index >= size) {
            return;
        }
        
        //因为有虚拟头节点，所以不用对index=0的情况进行特殊处理
        ListNode pre = head;
        for (int i = 0; i < index ; i++) {
            pre = pre.next;
        }
        pre.next = pre.next.next;
        size--;
    }

    // ===================== 测试主函数 =====================
    public static void main(String[] args) {
        dingyilianbiao list = new dingyilianbiao();

        // 测试添加
        list.addAtHead(1);    // 链表：1
        list.addAtTail(3);    // 链表：1 -> 3
        list.addAtIndex(1,2); // 链表：1 -> 2 -> 3

        System.out.println("get(1): " + list.get(1)); // 应该输出 2

        // 测试删除
        list.deleteAtIndex(1); // 链表：1 -> 3
        System.out.println("get(1): " + list.get(1)); // 应该输出 3

        // 测试越界
        System.out.println("get(2): " + list.get(2)); // 应该输出 -1
    }
}

File: LinkedList/dingyilianbiao.cpp
#include <iostream>
using namespace std;

class MyLinkedList {
public:
    struct LinkNode
    {
        int val;
        LinkNode* next;
        LinkNode(int val)
        {
            this->val = val;
            this->next = nullptr;
        }
    };
    MyLinkedList() {
        dummyHead = new LinkNode(0);
        size = 0;
    }
    
    int get(int index) {
        if(index >= size || index < 0)
        {
            return -1;
        }
        LinkNode* cur = dummyHead->next;
        for(int i=0;i<index;i++)
        {
            cur = cur->next;
        }
        return cur->val;
    }
    
    void addAtHead(int val) {
        LinkNode* temp = new LinkNode(val);
        temp->next = dummyHead->next;
        dummyHead->next = temp;
        size++;
    }
    
    void addAtTail(int val) {
        LinkNode* temp = new LinkNode(val);
        LinkNode* cur = dummyHead;
        while(cur->next != nullptr)
        {
            cur = cur->next;
        }
        cur->next = temp;
        size++;
    }
    
    void addAtIndex(int index, int val) {
        if(index > size) return;
        if(index < 0) index = 0;       
        LinkNode* temp = new LinkNode(val);
        // 修正：cur 从 dummyHead 开始
        LinkNode* cur = dummyHead;
        for(int i=0;i<index;i++)
        {
            cur = cur->next;
        }
        temp->next = cur->next;
        cur->next = temp;
        size++;
    }
    
    void deleteAtIndex(int index) {
        if (index >= size || index < 0) {
            return;
        }
        LinkNode* cur = dummyHead;
        while(index--) {
            cur = cur->next;
        }
        LinkNode* tmp = cur->next;
        cur->next = cur->next->next;
        delete tmp;
        tmp = nullptr;
        size--;
    }

    // 辅助函数：打印链表
    void printList() {
        LinkNode* cur = dummyHead->next;
        while(cur) {
            cout << cur->val << " -> ";
            cur = cur->next;
        }
        cout << "null" << endl;
    }

private:
    int size;
    LinkNode* dummyHead;
};

// 测试用例
int main() {
    MyLinkedList* obj = new MyLinkedList();
    obj->addAtHead(1);
    obj->addAtTail(3);
    obj->addAtIndex(1, 2);  // 链表变为 1->2->3
    cout << "get(1): " << obj->get(1) << endl; // 输出 2
    obj->deleteAtIndex(1);  // 链表变为 1->3
    cout << "get(1): " << obj->get(1) << endl; // 输出 3
    obj->printList();
    delete obj;
    return 0;
}